File: C++/mutex.cpp
#include <stdio.h>
#include <semaphore.h>
#include <thread>
#include <vector>

using std::vector, std::thread;

sem_t mutex;
int counter = 0;
vector<thread> threads;

void count() {
    sem_wait(&mutex);
    counter = counter + 1;
    printf("counter: %d\n", counter);
    sem_post(&mutex);
}

int main() {
    sem_init(&mutex, 0, 1);

    for (size_t i = 0; i < 20; ++i) {
        threads.push_back(thread(count));
    }

    for (auto &t : threads) {
        t.join();
    }

    sem_destroy(&mutex);

    return 0;
}

File: python/mutex.py
from threading import *

mutex = Semaphore(1)
counter = 0
threads = []

def count():
    global counter
    
    mutex.acquire()
    counter = counter + 1
    print(counter)
    mutex.release()

for i in range(20):
    threads.append(Thread(target=count))

for thread in threads:
    thread.start()

for thread in threads:
    thread.join()
